File: tianshu/core/skills/repository.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

import yaml


@dataclass
class Skill:
    name: str
    description: str
    instructions: str
    path: Path
    files: dict[str, str] = field(default_factory=dict)


def _parse_frontmatter(text: str) -> tuple[dict, str]:
    if not text.startswith("---"):
        return {}, text
    end = text.find("---", 3)
    if end == -1:
        return {}, text
    meta = text[3:end]
    body = text[end + 3 :]
    try:
        data = yaml.safe_load(meta) or {}
    except yaml.YAMLError:
        data = {}
    return data, body.lstrip("\n")


class SkillRepository:
    def __init__(self, root: Path) -> None:
        self.root = root
        self._skills: dict[str, Skill] = {}

    def scan(self) -> None:
        self._skills.clear()
        if not self.root.exists():
            return
        for skill_dir in self.root.iterdir():
            if not skill_dir.is_dir():
                continue
            md = skill_dir / "SKILL.md"
            if not md.exists():
                continue
            meta, body = _parse_frontmatter(md.read_text(encoding="utf-8"))
            name = meta.get("name") or skill_dir.name
            files: dict[str, str] = {}
            for f in skill_dir.iterdir():
                if f.name == "SKILL.md":
                    continue
                files[f.name] = f.read_text(encoding="utf-8")
            self._skills[name] = Skill(
                name=name,
                description=meta.get("description", ""),
                instructions=body,
                path=skill_dir,
                files=files,
            )

    def get(self, name: str) -> Skill | None:
        return self._skills.get(name)

    def list(self) -> list[Skill]:
        return list(self._skills.values())

    def descriptions(self) -> str:
        lines = []
        for s in self._skills.values():
            lines.append(f"- {s.name}: {s.description}")
        return "\n".join(lines)

File: tiankui-ts/src/core/skills/repository.ts
import * as fs from "node:fs";
import * as path from "node:path";

export interface Skill {
  name: string;
  description: string;
  instructions: string;
  path: string;
  files: Record<string, string>;
}

export function parseFrontmatter(text: string): { meta: Record<string, unknown>; body: string } {
  if (!text.startsWith("---")) return { meta: {}, body: text };
  const end = text.indexOf("---", 3);
  if (end === -1) return { meta: {}, body: text };
  const metaText = text.slice(3, end);
  const body = text.slice(end + 3).replace(/^\n+/, "");
  const meta: Record<string, unknown> = {};
  for (const line of metaText.split("\n")) {
    const m = line.match(/^([^:]+):\s*(.*)$/);
    if (m) meta[m[1].trim()] = m[2].trim();
  }
  return { meta, body };
}

export class SkillRepository {
  root: string;
  private _skills = new Map<string, Skill>();

  constructor(root: string) {
    this.root = root;
  }

  scan(): void {
    this._skills.clear();
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(this.root, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const mdPath = path.join(this.root, entry.name, "SKILL.md");
      if (!fs.existsSync(mdPath)) continue;
      const text = fs.readFileSync(mdPath, "utf-8");
      const { meta, body } = parseFrontmatter(text);
      const name = String(meta.name ?? entry.name);
      const files: Record<string, string> = {};
      for (const f of fs.readdirSync(path.join(this.root, entry.name))) {
        if (f === "SKILL.md") continue;
        files[f] = fs.readFileSync(path.join(this.root, entry.name, f), "utf-8");
      }
      this._skills.set(name, {
        name,
        description: String(meta.description ?? ""),
        instructions: body,
        path: path.join(this.root, entry.name),
        files,
      });
    }
  }

  get(name: string): Skill | null {
    return this._skills.get(name) ?? null;
  }

  list(): Skill[] {
    return [...this._skills.values()];
  }

  descriptions(): string {
    return this.list()
      .map((s) => `- ${s.name}: ${s.description}`)
      .join("\n");
  }
}
